libraries/Microsoft.Bot.Builder/BotState.cs

```diff
@@ -48,84 +118,109 @@ public BotState(IStorage storage, string propertyName, Func<ITurnContext, string
             }
+        }
 
-            var key = _keyDelegate(context);
-
-            changes.Add(key, state);
-
-            if (_settings.LastWriterWins)
+        /// <summary>
+        /// Reset the state object to it's default form
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public Task ClearStateAsync(ITurnContext context, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var cachedState = context.Services.Get<CachedBotState>(this.contextServiceKey);
+            if (cachedState != null)
             {
-                foreach (var item in changes)
-                {
-                    if (item.Value is IStoreItem valueStoreItem)
-                    {
-                        valueStoreItem.ETag = "*";
-                    }
-                }
+                context.Services[this.contextServiceKey] = new CachedBotState();
             }
-
-            await _storage.WriteAsync(changes, cancellationToken).ConfigureAwait(false);
+            return Task.CompletedTask;
         }
 
-        protected virtual async Task ReadToContextServiceAsync(ITurnContext context, CancellationToken cancellationToken)
+        /// <summary>
+        /// IPropertyContainer.GetAsync() method gives IPropertyAccessor ability to get from it's container
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="turnContext"></param>
+        /// <param name="propertyName"></param>
+        /// <returns></returns>
+        public Task<T> GetPropertyAsync<T>(ITurnContext turnContext, string propertyName)
         {
-            var key = _keyDelegate(context);
-            var items = await _storage.ReadAsync(new[] { key }, cancellationToken).ConfigureAwait(false);
-            var state = items.Where(entry => entry.Key == key).Select(entry => entry.Value).OfType<TState>().FirstOrDefault();
-            if (state == null)
-            {
-                state = new TState();
-            }
+            var cachedState = turnContext.Services.Get<CachedBotState>(this.contextServiceKey);
+            if (cachedState.State.TryGetValue(propertyName, out object result))
+                return Task.FromResult<T>((T)result);
+            return Task.FromResult(default(T));
+        }
 
-            context.Services.Add(_propertyName, state);
+        /// <summary>
+        /// IPropertyContainer.DeleteAsync() method gives IPropertyAccessor ability to delete from it's container
+        /// </summary>
+        /// <param name="turnContext"></param>
+        /// <param name="propertyName"></param>
+        /// <returns></returns>
+        public Task DeletePropertyAsync(ITurnContext turnContext, string propertyName)
+        {
+            var cachedState = turnContext.Services.Get<CachedBotState>(this.contextServiceKey);
+            cachedState.State.Remove(propertyName);
```

Comment:
cachedState could be null here and you'll end up with a NullReferenceException calling .State . Add null check... and a unit test to cover the scenario.